### PR TITLE
Fix netflow v9

### PR DESCRIPTION
**Pull request by gpotter2**
Reworking of the NetflowV9 implementation (by adding a `netflowv9_defragment` mechanic) function, to fix
https://github.com/secdev/scapy/issues/916 and https://github.com/secdev/scapy/issues/1060.

Indeed, some NetflowV9 frames needs other, previously recieved frames, to be dissected.

### CODE REVIEW

**Comment by p-l-** [comment]
Hi,

Thanks for this fix. I don't really like the addition to `PacketList` (and `Packet`) just for this protocol (particularly undocumented).

I'll try to write a PoC and you'll tell me what you think.

**Comment by gpotter2** [comment]
Hi,
I could replace the plist_post_dissect functions by a function which one would pass the plist to.
Would it be better ?

**Comment by gpotter2** [comment]
@p-l- Hi, I've removed all the Plist/packet implementation.

Now, one can just call
`netflowv9_defragment(plist)`
to process the packet list

**Comment by p-l-** [comment]
I'm writing a "stateful" `PacketList` object that could be useful here. But for now, I'd like to have a Python 3 compatible release before working harder on it.

**Comment by gpotter2** [comment]
@p-l- "stateful" PacketList is a great idea,
However, i would like having working code in the release (current netflowv9 integration is not working).

This fix should be included in 2.4, and `netflowv9_defragment` will do the trick until we get the stateful packetlist.

What do you think, could you review it ? Thanks !

File: scapy/layers/netflow.py [hunk 1]
```diff
@@ -341,45 +383,87 @@ def dispatch_hook(cls, _pkt=None, *args, **kargs):
...
         return cls
-    
-    def post_dissection(self, pkt):
-        # We need the whole packet to be dissected to access field def in NetflowFlowsetV9
+
+def netflowv9_defragment(plist):
+    """Process all NetflowV9 Packets to match IDs of the DataFlowsets with the Headers.
+    plist: the list of mixed NetflowV9 packets."""
+    # We need the whole packet to be dissected to access field def in NetflowFlowsetV9 or NetflowOptionsFlowsetV9
+    packet_list = plist.filter(lambda x: x.haslayer(NetflowFlowsetV9) or x.haslayer(NetflowOptionsFlowsetV9))
```
Comment (p-l-) [inline scapy/layers/netflow.py]: Don't use `.filter()` + `lambda`, prefer something like `packet_list = [pkt for pkt in plist if [...]]`

File: scapy/layers/netflow.py [hunk 2]
```diff
@@ -341,45 +383,87 @@ def dispatch_hook(cls, _pkt=None, *args, **kargs):
...
+                        break
+            if NetflowOptionsFlowsetV9 in p: ## STEP 2 - NetflowOptionsFlowsetV9
+                current = p[NetflowOptionsFlowsetV9]
+                current_ftl = root.getlayer(NetflowDataflowsetV9, templateID=current.templateID)
                 if current_ftl:
                     # Matched
-                    if len(current_ftl.records) > 1:
-                        # post_dissection is not necessary
-                        return
+                    if len(current_ftl.records) < 0 or (not hasattr(current_ftl.records[0], "fieldValue")):
```
Comment (p-l-) [inline scapy/layers/netflow.py]: Maybe you could `try...except` here rather than using `hashattr()`

File: scapy/layers/netflow.py (hunk 1 above)
Comment (gpotter2) [reply scapy/layers/netflow.py]: Updated

File: scapy/layers/netflow.py (hunk 2 above)
Comment (gpotter2) [reply scapy/layers/netflow.py]: Good idea

**Comment by gpotter2** [comment]
Hi @p-l-, have you made progress in your “stateful” packetlist ?

Could it be used to handle sessions (such as this NetflowV9 one) ?

It would be cool to get rid of `netflowv9_defragment`, which was a “temporary” function intended to have Netflow v9 working for 2.4.0.

**Pull request merged**